ganga/GangaCore/GPIDev/Base/CoW.py

```diff
@@ -0,0 +1,504 @@
+#
+# Base concept taken from Yannick Loiseau
+# http://yloiseau.net/articles/DesignPatterns/flyweight/
+#
+
+from __future__ import with_statement
+from __future__ import absolute_import
+import weakref
+from copy import copy
+import types
+from collections import OrderedDict
+from operator import itemgetter
+import gc
+
+# Thing to explicitly not try to flyweight
+flyweight_ignored_keys = [
+                          "_flyweight_cache",
+                          "_flyweight_cb_func",
+                          "_my_flyweight_cb_func"
+                        ]
+flyweight_ignored_types = (int, float, type(None), bool, str) + weakref.ProxyTypes
+
+
+def proxify(value):
+    """
+        Wrap the value in a proxy shell if need be.
+        Returns the object or the proxy object.
+    """
+    if type(value) is unicode:
+        value = ProxyStr(value)
+
+    elif type(value) is list:
+        value = ProxyList(value)
+
+    elif type(value) is tuple:
+        value = ProxyTuple(value)
+
+    elif type(value) is set:
+        value = ProxySet(value)
+
+    elif type(value) is dict:
+        value = ProxyDict(value)
+
+    return value
+
+
+class CoW(object):
+
+    """
+        This class is a base class.
+        Traditional FlyWeight pattern is useful for ensuring we
+        do not duplicate memory. However, given  we also wanted
+        to utilize Copy-On-Write principles for performance.
+        Thus, we take the concept of FlyWeight pattern, and
+        extend it by instrumenting getattr and setattr. The net
+        result is that, transparent to the underlying python class object,
+        every attribute is actually a pointer and memory is reduced in that
+        manner. Also, I override the __copy__ method to copy the pointers
+        into a new object instead of fully traversing get/set.
+        This allows me to not have to implement
+        custom copy methods per each class, and instead utilize the
+        duck typing of python to make this all transparent.
+    """
+
+    # __slots__ = ['__weakref__','__cow_add_cb_pointer','_my_flyweight_cb_func','_flyweight_cache']
+
+    # Using weakref to allow garbage collection
+    # dict[<var_type>][__hash__]
+    _flyweight_cache = {}
+
+    def __init__(self, *args, **kwargs):
```

Comment:
Could you also define some comparison operators as well? It might be useful when this gets integrated into python3

Reply:
yes, I am working on it. also while switching to python3 we could use `OrderedDict` instead of `dict` in ProxyDict, we won't have to sort the dictionary for getting hash keys then.

Reply:
If you feel that the `OrderedDict` approach is better and should be what we are using for python 3 then once this code is merged into the python 3 branch please make the change and open a PR against the python 3 branch.

Reply:
Yes.. Sure @alexanderrichards